File: frontend/src/api/unifiedGraph.ts
import apiClient from './client'
import { useQuery } from '@tanstack/react-query'

// ============================================================================
// Types
// ============================================================================

export interface GraphNodeData {
  id: string
  label: string
  node_type: string
  type?: string
  canonical_key?: string
  district?: string
  province?: string
  latitude?: number
  longitude?: number
  subtype?: string
  confidence: number
  properties: Record<string, unknown>
  // Metrics (when available)
  degree?: number
  pagerank?: number
  is_hub?: boolean
  is_bridge?: boolean
  // Compound node fields
  member_count?: number
  parent?: string
}

export interface GraphNode {
  data: GraphNodeData
}

export interface GraphEdgeData {
  id: string
  source: string
  target: string
  predicate: string
  weight: number
  confidence: number
  is_current: boolean
  properties: Record<string, unknown>
}

export interface GraphEdge {
  data: GraphEdgeData
}

export interface DistrictSummary {
  id: string
  name_en: string
  name_ne?: string
  province_id: number
  province_name: string
  graph_node_id?: string
  node_count: number
  latitude?: number
  longitude?: number
}

export interface GraphOverviewResponse {
  nodes: GraphNode[]
  edges: GraphEdge[]
  rendered_nodes: number
  rendered_edges: number
  total_graph_nodes: number
  total_graph_edges: number
}

export interface GraphExpandResponse {
  nodes: GraphNode[]
  edges: GraphEdge[]
  total_neighbors: number
  has_more: boolean
  offset: number
  limit: number
}

export interface GraphNeighborhoodResponse {
  nodes: GraphNode[]
  edges: GraphEdge[]
  depth_reached: number
}

export interface GraphPathResponse {
  found: boolean
  path: string[]      // node UUIDs in traversal order
  edges: string[]     // edge UUIDs along the path
  length: number
}

export interface GraphSearchResponse {
  nodes: GraphNode[]
  total: number
}

export interface GraphStatsResponse {
  total_nodes: number
  total_edges: number
  by_type: Record<string, number>
  by_predicate: Record<string, number>
  by_district: Record<string, number>
}

export interface GraphHealthDomainCoverage {
  source_table: string
  total_nodes: number
  connected_nodes: number
  coverage_ratio: number
}

export interface GraphHealthResponse {
  status: 'healthy' | 'degraded' | string
  total_nodes: number
  total_edges: number
  connected_node_ratio: number
  largest_component_ratio: number
  per_domain_coverage: GraphHealthDomainCoverage[]
  thresholds_breached: string[]
}

export interface GraphTimeseriesBucket {
  bucket_start: string
  total_edges: number
  by_predicate: Record<string, number>
  by_domain: Record<string, number>
}

export interface GraphTimeseriesResponse {
  window: string
  bucket: 'hour' | 'day' | 'week' | string
  from_ts?: string
  to_ts?: string
  series: GraphTimeseriesBucket[]
}

export interface EdgeDetail {
  id: string
  predicate: string
  direction: 'outgoing' | 'incoming'
  peer_id: string
  peer_title?: string
  peer_type?: string
  weight: number
  confidence: number
  valid_from?: string
  valid_to?: string
  properties: Record<string, unknown>
}

export interface GraphNodeDetailResponse {
  id: string
  node_type: string
  type?: string
  canonical_key?: string
  title: string
  title_ne?: string
  subtitle?: string
  description?: string
  subtype?: string
  district?: string
  province?: string
  latitude?: number
  longitude?: number
  properties: Record<string, unknown>
  source_table?: string
  source_id?: string
  confidence: number
  source_count: number
  is_canonical: boolean
  canonical_node_id?: string
  first_seen_at?: string
  last_seen_at?: string
  created_at?: string
  edges: EdgeDetail[]
  total_outgoing: number
  total_incoming: number
  metrics: Record<string, unknown>[]
  resolutions: Record<string, unknown>[]
}

export interface DistrictListResponse {
  districts: DistrictSummary[]
  total: number
}

export interface ResolveNodeRequest {
  source_table: string
  source_id: string
  canonical_key?: string
}

export interface ResolveNodeResponse {
  found: boolean
  node?: GraphNode
}

export interface UnifiedTimelineEvent {
  event_type: string
  timestamp?: string
  title: string
  object_id?: string
  link_id?: string
  confidence?: number
  source_count?: number
  verification_status?: string
  provenance_refs?: Record<string, unknown>[]
}

export interface UnifiedTimelineResponse {
  center: {
    id: string
    title: string
    object_type: string
  }
  events: UnifiedTimelineEvent[]
  total: number
}

export interface NodeProfileResponse {
  node: {
    id: string
    title: string
    node_type: string
    type?: string
    subtype?: string
    district?: string
    province?: string
    latitude?: number
    longitude?: number
    description?: string
    source_table?: string
    source_id?: string
    properties: Record<string, unknown>
    confidence: number
  }
  profile_type: string
  summary: Record<string, unknown>
  relationships: {
    total: number
    by_predicate: Record<string, number>
    top_neighbors: Array<{
      peer_id?: string
      peer_title?: string
      peer_type?: string
      predicate?: string
      confidence?: number
    }>
  }
  quality: {
    quality_score: number
    missing_fields: string[]
    provenance_count: number
    last_updated?: string
  }
}

export interface GraphCorrectionEntry {
  id: string
  action: string
  status: 'pending' | 'approved' | 'rejected' | 'rolled_back' | string
  node_id?: string
  edge_id?: string
  payload: Record<string, unknown>
  reason: string
  submitted_by: string
  submitted_by_email: string
  submitted_at?: string
  reviewed_by?: string
  reviewed_at?: string
  review_notes?: string
  rejection_reason?: string
  applied_change?: Record<string, unknown>
  applied_at?: string
  rolled_back_by?: string
  rolled_back_at?: string
  rollback_reason?: string
}

export interface GraphCorrectionListResponse {
  items: GraphCorrectionEntry[]
  pending_count: number
  page: number
  total: number
  total_pages: number
}

export interface GraphCorrectionActionResponse {
  id: string
  status: string
  message: string
}

// ============================================================================
// Query parameter interfaces
// ============================================================================

export interface ExpandNodeParams {
  offset?: number
  limit?: number
  predicates?: string
  min_confidence?: number
  as_of?: string
  from_ts?: string
  to_ts?: string
  window?: '24h' | '7d' | '30d' | 'all_time'
  include_inferred?: boolean
}

export interface NeighborhoodParams {
  depth?: number
  limit?: number
  min_confidence?: number
  node_types?: string
  as_of?: string
  from_ts?: string
  to_ts?: string
  window?: '24h' | '7d' | '30d' | 'all_time'
  include_inferred?: boolean
}

export interface SearchParams {
  node_types?: string
  districts?: string
  limit?: number
}

export interface TemporalQueryParams {
  as_of?: string
  from_ts?: string
  to_ts?: string
  window?: '24h' | '7d' | '30d' | 'all_time'
  include_inferred?: boolean
}

// ============================================================================
// API functions
// ============================================================================

const GRAPH_BASE = '/unified-graph'

export const getGraphOverview = (params?: TemporalQueryParams) =>
  apiClient.get<GraphOverviewResponse>(`${GRAPH_BASE}/overview`, { params })

export const expandNode = (nodeId: string, params?: ExpandNodeParams) =>
  apiClient.get<GraphExpandResponse>(`${GRAPH_BASE}/expand/${nodeId}`, { params })

export const getNeighborhood = (nodeId: string, params?: NeighborhoodParams) =>
  apiClient.get<GraphNeighborhoodResponse>(`${GRAPH_BASE}/neighborhood/${nodeId}`, { params })

export const findPath = (fromNode: string, toNode: string, maxDepth?: number) =>
  apiClient.get<GraphPathResponse>(`${GRAPH_BASE}/path`, {
    params: { from_node: fromNode, to_node: toNode, max_depth: maxDepth },
  })

export const searchGraph = (q: string, params?: SearchParams) =>
  apiClient.get<GraphSearchResponse>(`${GRAPH_BASE}/search`, {
    params: { q, ...params },
  })

export const getGraphStats = (params?: TemporalQueryParams) =>
  apiClient.get<GraphStatsResponse>(`${GRAPH_BASE}/stats`, { params })

export const getGraphHealth = () =>
  apiClient.get<GraphHealthResponse>(`${GRAPH_BASE}/health`)

export const getGraphTimeseries = (params?: {
  bucket?: 'hour' | 'day' | 'week'
} & TemporalQueryParams) =>
  apiClient.get<GraphTimeseriesResponse>(`${GRAPH_BASE}/timeseries`, { params })

export const getNodeDetail = (nodeId: string, params?: TemporalQueryParams) =>
  apiClient.get<GraphNodeDetailResponse>(`${GRAPH_BASE}/node/${nodeId}`, { params })

export const listDistricts = () =>
  apiClient.get<DistrictListResponse>(`${GRAPH_BASE}/districts`)

export const resolveGraphNode = (payload: ResolveNodeRequest) =>
  apiClient.post<ResolveNodeResponse>(`${GRAPH_BASE}/resolve-node`, payload)

export const getUnifiedNodeTimeline = (nodeId: string, limit = 100) =>
  apiClient.get<UnifiedTimelineResponse>(`${GRAPH_BASE}/node/${nodeId}/timeline`, {
    params: { limit },
  })

export const getUnifiedNodeProfile = (nodeId: string) =>
  apiClient.get<NodeProfileResponse>(`${GRAPH_BASE}/node/${nodeId}/profile`)

export const submitGraphCorrection = (payload: {
  action: string
  reason: string
  payload: Record<string, unknown>
  node_id?: string
  edge_id?: string
}) => apiClient.post<GraphCorrectionActionResponse>(`${GRAPH_BASE}/corrections`, payload)

export const fetchGraphCorrections = (params?: {
  status?: string
  page?: number
  per_page?: number
}) => apiClient.get<GraphCorrectionListResponse>(`${GRAPH_BASE}/corrections`, { params })

export const approveGraphCorrection = (correctionId: string, notes?: string) =>
  apiClient.post<GraphCorrectionActionResponse>(
    `${GRAPH_BASE}/corrections/${correctionId}/approve`,
    null,
    { params: { notes } },
  )

export const rejectGraphCorrection = (correctionId: string, reason: string) =>
  apiClient.post<GraphCorrectionActionResponse>(
    `${GRAPH_BASE}/corrections/${correctionId}/reject`,
    null,
    { params: { reason } },
  )

export const rollbackGraphCorrection = (correctionId: string, reason: string) =>
  apiClient.post<GraphCorrectionActionResponse>(
    `${GRAPH_BASE}/corrections/${correctionId}/rollback`,
    null,
    { params: { reason } },
  )

// ============================================================================
// React Query hooks
// ============================================================================

export const useGraphOverview = () =>
  useQuery({
    queryKey: ['unified-graph-overview'],
    queryFn: () => getGraphOverview().then((r) => r.data),
    staleTime: 120_000,
  })

export const useExpandNode = (nodeId: string | null, params?: ExpandNodeParams) =>
  useQuery({
    queryKey: ['unified-graph-expand', nodeId, params],
    queryFn: () => expandNode(nodeId!, params).then((r) => r.data),
    enabled: !!nodeId,
    staleTime: 60_000,
  })

export const useNeighborhood = (nodeId: string | null, params?: NeighborhoodParams) =>
  useQuery({
    queryKey: ['unified-graph-neighborhood', nodeId, params],
    queryFn: () => getNeighborhood(nodeId!, params).then((r) => r.data),
    enabled: !!nodeId,
    staleTime: 60_000,
  })

export const useGraphSearch = (query: string, params?: SearchParams) =>
  useQuery({
    queryKey: ['unified-graph-search', query, params],
    queryFn: () => searchGraph(query, params).then((r) => r.data),
    enabled: query.length >= 2,
    staleTime: 30_000,
  })

export const useGraphStats = () =>
  useQuery({
    queryKey: ['unified-graph-stats'],
    queryFn: () => getGraphStats().then((r) => r.data),
    staleTime: 300_000,
  })

export const useGraphHealth = () =>
  useQuery({
    queryKey: ['unified-graph-health'],
    queryFn: () => getGraphHealth().then((r) => r.data),
    staleTime: 60_000,
  })

export const useNodeDetail = (nodeId: string | null) =>
  useQuery({
    queryKey: ['unified-graph-node', nodeId],
    queryFn: () => getNodeDetail(nodeId!).then((r) => r.data),
    enabled: !!nodeId,
    staleTime: 60_000,
  })

export const useDistricts = () =>
  useQuery({
    queryKey: ['unified-graph-districts'],
    queryFn: () => listDistricts().then((r) => r.data),
    staleTime: 600_000,
  })

File: backend-v5/app/schemas/unified_graph.py
"""Pydantic schemas for Unified Graph API endpoints."""
from datetime import datetime
from typing import Optional, Dict, List, Any
from uuid import UUID

from pydantic import BaseModel, Field


# ============================================================
# Core Graph Element Schemas
# ============================================================

class GraphNodeResponse(BaseModel):
    """Full graph node detail from the database."""
    id: UUID
    node_type: str
    canonical_key: str
    title: str
    title_ne: Optional[str] = None
    subtitle: Optional[str] = None
    district: Optional[str] = None
    province: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    properties: Dict[str, Any] = Field(default_factory=dict)
    confidence: float = 1.0
    is_canonical: bool = True
    source_table: str


class GraphEdgeResponse(BaseModel):
    """Full graph edge detail from the database."""
    id: UUID
    source_node_id: UUID
    target_node_id: UUID
    predicate: str
    weight: float = 1.0
    confidence: float = 1.0
    is_current: bool = True
    valid_from: Optional[datetime] = None
    valid_to: Optional[datetime] = None
    properties: Dict[str, Any] = Field(default_factory=dict)


# ============================================================
# Cytoscape.js Compatible Formats
# ============================================================

class CytoscapeNode(BaseModel):
    """Cytoscape.js compatible node format.

    data dict contains: id, label, node_type, district, province,
    latitude, longitude, confidence, member_count, and all
    type-specific properties from the graph_node.properties JSONB.
    """
    data: Dict[str, Any]


class CytoscapeEdge(BaseModel):
    """Cytoscape.js compatible edge format.

    data dict contains: id, source, target, predicate, weight,
    confidence, is_current, valid_from, valid_to, and all
    edge-specific properties from graph_edges.properties JSONB.
    """
    data: Dict[str, Any]


# ============================================================
# API Response Schemas
# ============================================================

class GraphOverviewResponse(BaseModel):
    """Response for GET /unified-graph/overview.

    Returns all 77 district + 7 province nodes with inter-district
    edges. Includes total graph counts for status bar display.
    """
    nodes: List[CytoscapeNode]
    edges: List[CytoscapeEdge]
    rendered_nodes: int = 0
    rendered_edges: int = 0
    total_graph_nodes: int
    total_graph_edges: int


class GraphExpandResponse(BaseModel):
    """Response for GET /unified-graph/expand/{node_id}.

    Returns direct neighbors of a node with pagination metadata.
    """
    nodes: List[CytoscapeNode]
    edges: List[CytoscapeEdge]
    total_neighbors: int
    has_more: bool
    offset: int = 0
    limit: int = 50


class GraphNeighborhoodResponse(BaseModel):
    """Response for GET /unified-graph/neighborhood/{node_id}.

    Returns all nodes and edges within `depth` hops of a node.
    """
    nodes: List[CytoscapeNode]
    edges: List[CytoscapeEdge]
    depth_reached: int


class GraphPathResponse(BaseModel):
    """Response for GET /unified-graph/path.

    Returns the shortest path between two nodes as ordered UUID lists.
    ``path`` contains node IDs in traversal order.
    ``edges`` contains edge IDs along the path.
    """
    found: bool
    path: List[str] = Field(default_factory=list)
    edges: List[str] = Field(default_factory=list)
    length: int = 0


class GraphSearchResponse(BaseModel):
    """Response for GET /unified-graph/search.

    Returns matching nodes from full-text search with total count.
    """
    nodes: List[CytoscapeNode]
    total: int


class GraphStatsResponse(BaseModel):
    """Response for GET /unified-graph/stats.

    Aggregated counts for dashboard / status bar display.
    """
    total_nodes: int
    total_edges: int
    by_type: Dict[str, int] = Field(default_factory=dict)
    by_predicate: Dict[str, int] = Field(default_factory=dict)
    by_district: Dict[str, int] = Field(default_factory=dict)


class GraphHealthDomainCoverage(BaseModel):
    source_table: str
    total_nodes: int
    connected_nodes: int
    coverage_ratio: float


class GraphHealthResponse(BaseModel):
    status: str
    total_nodes: int
    total_edges: int
    connected_node_ratio: float
    largest_component_ratio: float
    per_domain_coverage: List[GraphHealthDomainCoverage] = Field(default_factory=list)
    thresholds_breached: List[str] = Field(default_factory=list)


class GraphTimeseriesBucket(BaseModel):
    bucket_start: str
    total_edges: int
    by_predicate: Dict[str, int] = Field(default_factory=dict)
    by_domain: Dict[str, int] = Field(default_factory=dict)


class GraphTimeseriesResponse(BaseModel):
    window: str
    bucket: str
    from_ts: Optional[str] = None
    to_ts: Optional[str] = None
    series: List[GraphTimeseriesBucket] = Field(default_factory=list)


class NodeDetailResponse(BaseModel):
    """Response for GET /unified-graph/node/{node_id}.

    Full detail for a single node as a flat dict including its edges,
    metrics, and resolution history.
    """
    id: str
    node_type: str
    type: Optional[str] = None
    canonical_key: Optional[str] = None
    title: str
    title_ne: Optional[str] = None
    subtitle: Optional[str] = None
    description: Optional[str] = None
    subtype: Optional[str] = None
    district: Optional[str] = None
    province: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    properties: Dict[str, Any] = Field(default_factory=dict)
    source_table: Optional[str] = None
    source_id: Optional[str] = None
    confidence: float = 1.0
    source_count: int = 1
    is_canonical: bool = True
    canonical_node_id: Optional[str] = None
    first_seen_at: Optional[str] = None
    last_seen_at: Optional[str] = None
    created_at: Optional[str] = None
    edges: List[Dict[str, Any]] = Field(default_factory=list)
    total_outgoing: int = 0
    total_incoming: int = 0
    temporal_profile: Dict[str, Any] = Field(default_factory=dict)
    metrics: List[Dict[str, Any]] = Field(default_factory=list)
    resolutions: List[Dict[str, Any]] = Field(default_factory=list)


class DistrictResponse(BaseModel):
    """District reference data for filter dropdowns."""
    id: UUID
    name_en: str
    name_ne: Optional[str] = None
    province_id: int
    province_name: str
    graph_node_id: Optional[UUID] = None
    node_count: int = 0
    latitude: Optional[float] = None
    longitude: Optional[float] = None


class DistrictListResponse(BaseModel):
    """Response for GET /unified-graph/districts."""
    districts: List[DistrictResponse]
    total: int


class ResolveNodeRequest(BaseModel):
    """Request for POST /unified-graph/resolve-node."""
    source_table: str
    source_id: str
    canonical_key: Optional[str] = None


class ResolveNodeResponse(BaseModel):
    """Response for POST /unified-graph/resolve-node."""
    found: bool
    node: Optional[CytoscapeNode] = None


class UnifiedTimelineCenter(BaseModel):
    id: str
    title: str
    object_type: str


class UnifiedTimelineEvent(BaseModel):
    event_type: str
    timestamp: Optional[str] = None
    title: str
    object_id: Optional[str] = None
    link_id: Optional[str] = None
    confidence: Optional[float] = None
    source_count: Optional[int] = None
    verification_status: Optional[str] = None
    provenance_refs: List[Dict[str, Any]] = Field(default_factory=list)


class NodeTimelineResponse(BaseModel):
    """Response for GET /unified-graph/node/{node_id}/timeline."""
    center: UnifiedTimelineCenter
    events: List[UnifiedTimelineEvent] = Field(default_factory=list)
    total: int = 0


class NodeProfileNode(BaseModel):
    id: str
    title: str
    node_type: str
    type: Optional[str] = None
    subtype: Optional[str] = None
    district: Optional[str] = None
    province: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    description: Optional[str] = None
    source_table: Optional[str] = None
    source_id: Optional[str] = None
    properties: Dict[str, Any] = Field(default_factory=dict)
    confidence: float = 0.0


class NodeProfileQuality(BaseModel):
    quality_score: float = 0.0
    missing_fields: List[str] = Field(default_factory=list)
    provenance_count: int = 0
    last_updated: Optional[str] = None


class NodeProfileRelationships(BaseModel):
    total: int = 0
    by_predicate: Dict[str, int] = Field(default_factory=dict)
    top_neighbors: List[Dict[str, Any]] = Field(default_factory=list)


class NodeProfileResponse(BaseModel):
    """Response for GET /unified-graph/node/{node_id}/profile."""
    node: NodeProfileNode
    profile_type: str
    summary: Dict[str, Any] = Field(default_factory=dict)
    relationships: NodeProfileRelationships
    quality: NodeProfileQuality


class GraphCorrectionRequest(BaseModel):
    """Request body for POST /unified-graph/corrections."""
    action: str
    payload: Dict[str, Any] = Field(default_factory=dict)
    reason: str
    node_id: Optional[str] = None
    edge_id: Optional[str] = None


class GraphCorrectionEntry(BaseModel):
    id: str
    action: str
    status: str
    node_id: Optional[str] = None
    edge_id: Optional[str] = None
    payload: Dict[str, Any] = Field(default_factory=dict)
    reason: str
    submitted_by: str
    submitted_by_email: str = ""
    submitted_at: Optional[str] = None
    reviewed_by: Optional[str] = None
    reviewed_at: Optional[str] = None
    review_notes: Optional[str] = None
    rejection_reason: Optional[str] = None
    applied_change: Optional[Dict[str, Any]] = None
    applied_at: Optional[str] = None
    rolled_back_by: Optional[str] = None
    rolled_back_at: Optional[str] = None
    rollback_reason: Optional[str] = None


class GraphCorrectionListResponse(BaseModel):
    items: List[GraphCorrectionEntry] = Field(default_factory=list)
    pending_count: int = 0
    page: int = 1
    total: int = 0
    total_pages: int = 0


class GraphCorrectionActionResponse(BaseModel):
    id: str
    status: str
    message: str


# ============================================================
# Ingestion / Pipeline Response Schemas
# ============================================================

class PipelineJobResponse(BaseModel):
    """Response for POST endpoints that trigger background pipelines."""
    status: str = "accepted"
    message: str
    phases_requested: List[str] = Field(default_factory=list)
    run_id: Optional[str] = None
    results: Dict[str, Any] = Field(default_factory=dict)


class EntityResolutionResponse(BaseModel):
    """Response for POST /unified-graph/resolve."""
    status: str = "accepted"
    message: str
    method: str


class MetricsRecomputeResponse(BaseModel):
    """Response for POST /unified-graph/metrics/recompute."""
    status: str = "accepted"
    message: str
    window_type: str
